crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -413,9 +415,20 @@ impl TableMetadata {
         file_io: &FileIO,
         metadata_location: impl AsRef<str>,
     ) -> Result<TableMetadata> {
-        let input_file = file_io.new_input(metadata_location)?;
+        let input_file = file_io.new_input(metadata_location.as_ref())?;
         let metadata_content = input_file.read().await?;
-        let metadata = serde_json::from_slice::<TableMetadata>(&metadata_content)?;
+
+        let metadata = if metadata_location.as_ref().ends_with(".gz.metadata.json") {
```

Comment:
Do we want to optionally support the Java Iceberg alternative?

> The Java reference implementation can additionally read GZIP compressed files with the suffix metadata.json.gz.

Reply:
> Even better would be peeking at the file and looking for the gzip magic number. If there's interest in that I can implement it.

That would be a really elegant solution, I think.

Reply:
Ok, done!